File: caches/least_frequently_used_cache.py
class Node:
    def __init__(self, key, val, freq=1):
        self.key = key
        self.value = val
        self.cnt = 1
        self.prev = None
        self.next = None

class LFUCache:
    def __init__(self, capacity: int):
        self.cacheMap = {} 
        self.freqMap = {}
        self.minFreq = 0
        self.capacity = capacity
        
    def get(self, key: int) -> int:
        if key not in self.cacheMap:
            return -1
            
        node = self.cacheMap[key]
        res = node.value
        self.updateFreq(node)
        return res
        
    def put(self, key: int, value: int) -> None:
        if self.capacity == 0:
            return
            
        if key in self.cacheMap:
            node = self.cacheMap[key]
            node.value = value
            self.updateFreq(node)
            return

        else:
        
            if len(self.cacheMap) == self.capacity:

                node = self.freqMap[self.minFreq][1].prev
                del self.cacheMap[node.key]
                self.remove(node)
                
                if self.freqMap[self.minFreq][0].next == self.freqMap[self.minFreq][1]:

                    del self.freqMap[self.minFreq]

            node = Node(key, value)
            self.cacheMap[key] = node

            self.minFreq = 1
            self.add(node, 1)

    def add(self, node, freq):

        if freq not in self.freqMap:

            head = Node(-1, -1)
            
            tail = Node(-1, -1)
            head.next = tail
            tail.prev = head
            self.freqMap[freq] = (head, tail)

        head = self.freqMap[freq][0]
        temp = head.next
        node.next = temp
        node.prev = head
        head.next = node
        temp.prev = node

    def remove(self, node):
        delprev = node.prev
        delnext = node.next
        delprev.next = delnext
        delnext.prev = delprev

    def updateFreq(self, node):
        oldFreq = node.cnt

        node.cnt += 1

        self.remove(node)
        
        if self.freqMap[oldFreq][0].next == self.freqMap[oldFreq][1]:
            del self.freqMap[oldFreq]

            if self.minFreq == oldFreq:
                self.minFreq += 1

        self.add(node, node.cnt)




    def update(self, node):
        # Remove from old frequency list
        freq = node.freq
        self.freqMap[freq].remove(node)
        
        # Update minFreq if needed
        if self.minFreq == freq and self.freqMap[freq].size == 0:
            self.minFreq += 1
            
        # Add to new frequency list
        node.freq += 1
        if node.freq not in self.freqMap:
            self.freqMap[node.freq] = DLList()
        self.freqMap[node.freq].add(node)

File: caches/least_frequently_used_cache.rb
class Node
  attr_accessor :key, :value, :cnt, :prev, :next
  
  def initialize(key, val, freq = 1)
    @key = key
    @value = val
    @cnt = 1
    @prev = nil
    @next = nil
  end
end

class LFUCache
  def initialize(capacity)
    @cache_map = {}
    @freq_map = {}
    @min_freq = 0
    @capacity = capacity
  end

  def get(key)
    return -1 unless @cache_map.key?(key)
    
    node = @cache_map[key]
    res = node.value
    update_freq(node)
    res
  end

  def put(key, value)
    return if @capacity == 0
    
    if @cache_map.key?(key)
      node = @cache_map[key]
      node.value = value
      update_freq(node)
      return
    end

    if @cache_map.size == @capacity
      node = @freq_map[@min_freq][1].prev
      @cache_map.delete(node.key)
      remove(node)
      
      if @freq_map[@min_freq][0].next == @freq_map[@min_freq][1]
        @freq_map.delete(@min_freq)
      end
    end

    node = Node.new(key, value)
    @cache_map[key] = node
    @min_freq = 1
    add(node, 1)
  end

  private

  def add(node, freq)
    unless @freq_map.key?(freq)
      head = Node.new(-1, -1)
      tail = Node.new(-1, -1)
      head.next = tail
      tail.prev = head
      @freq_map[freq] = [head, tail]
    end

    head = @freq_map[freq][0]
    temp = head.next
    node.next = temp
    node.prev = head
    head.next = node
    temp.prev = node
  end

  def remove(node)
    del_prev = node.prev
    del_next = node.next
    del_prev.next = del_next
    del_next.prev = del_prev
  end

  def update_freq(node)
    old_freq = node.cnt
    node.cnt += 1
    remove(node)

    if @freq_map[old_freq][0].next == @freq_map[old_freq][1]
      @freq_map.delete(old_freq)
      @min_freq += 1 if @min_freq == old_freq
    end

    add(node, node.cnt)
  end
end